### PR TITLE
NPC Migration to using Powers

### PR DESCRIPTION
**Pull request by JFranklin3000**
##### In this PR
- Migration for Threat Powers to Powers

##### Testing
- create Threat Powers in items, compendiums, and on actors
- validate items migrate.

### CODE REVIEW

File: module/helpers/config.mjs [hunk 1]
```diff
@@ -326,11 +326,23 @@ E20.skillShiftList = [
 // Options for Actions
 E20.actionTypes = {
   free: "E20.ActionTypeFree",
+  fullAction: "E20.ActionTypeFullAction",
   move: "E20.ActionTypeMove",
   standard: "E20.ActionTypeStandard",
+  standardAndMove: "E20.ActionTypeStandardAndMove",
+  wholeTurn: "E20.ActionTypeWholeTurn",
+
 };
```
Comment (phildominguez) [inline module/helpers/config.mjs:335]:
```suggestion
  wholeTurn: "E20.ActionTypeWholeTurn",
};
```

File: module/migration.mjs [hunk 2]
```diff
@@ -35,6 +34,10 @@ export const migrateWorld = async function() {
   for (const [item, valid] of items) {
     try {
       const source = valid ? item.toObject() : game.data.items.find(i => i._id === item.id);
+      if (item.type == "threatPower") {
```
Comment (phildominguez) [inline module/migration.mjs:37]:
```suggestion
      const source = valid ? item.toObject() : game.data.items.find(i => i._id === item.id);

      if (item.type == "threatPower") {
```

File: module/migration.mjs [hunk 3]
```diff
@@ -203,10 +206,16 @@ export const migrateActorData = function(actor) {
...
   }
 
-  const items = actor.items.reduce((arr, i) => {
+  const items = actor.items.reduce(async (arr, i) => {
     // Migrate the Owned Item
     const itemData = i instanceof CONFIG.Item.documentClass ? i.toObject() : i;
-    let itemUpdate = migrateItemData(itemData);
+    const fullActor = game.actors.get(actor._id);
+    const itemToDelete = fullActor.items.get(i._id);
+    if (itemToDelete.type == "threatPower"){
```
Comment (phildominguez) [inline module/migration.mjs:214]:
```suggestion
    const itemToDelete = fullActor.items.get(i._id);

    if (itemToDelete.type == "threatPower") {
```

File: module/migration.mjs [hunk 4]
```diff
@@ -263,6 +272,29 @@ export function migrateItemData(item) {
...
       updateData[`system.type`] = perkType;
     }
+  } else if (item.type == "threatPower") {
+    const itemData = item;
+    itemData.type = "power";
+    itemData.system.canActivate = true;
+    itemData.system.usesPer = item.system.charges;
+    itemData.system.type = "threat";
+    itemData.system.usesInterval = "perScene";
+    //This is an attempt to catch as many actions as possible by converting to camelCase.
```
Comment (phildominguez) [inline module/migration.mjs:282]:
```suggestion
    itemData.system.usesInterval = "perScene";

    //This is an attempt to catch as many actions as possible by converting to camelCase.
```

File: module/sheet-handlers/power-handler.mjs [hunk 5]
```diff
@@ -78,6 +85,8 @@ export class PowerHandler {
       ).render(true);
     } else if (classFeature && classFeature.system.uses.value >= power.system.powerCost){
       classFeature.update({ ["system.uses.value"]: Math.max(0, classFeature.system.uses.value - power.system.powerCost) });
+    } else if (!power.system.powerCost) {
+      console.log("still working on something for here");
```
Comment (phildominguez) [inline module/sheet-handlers/power-handler.mjs:89]: 😲

**Review by phildominguez — CHANGES_REQUESTED** [review]

File: module/sheet-handlers/power-handler.mjs (hunk 5 above)
Comment (JFranklin3000) [reply module/sheet-handlers/power-handler.mjs:89]: That is a lint cheat :D

File: module/migration.mjs [hunk 6]
```diff
@@ -263,6 +274,30 @@ export function migrateItemData(item) {
...
+      itemData.system.actionType = "free";
+    }
+
+    if (actor) {
+      Item.implementation.create(itemData, {parent: actor, keepId: true});
+    } else {
+      Item.implementation.create(itemData, {keepId: true});
+    }
+
   }
```
Comment (phildominguez) [inline module/migration.mjs:301]:
```suggestion
    }
  }
```

**Review by phildominguez — CHANGES_REQUESTED** [review]

File: module/migration.mjs [hunk 7]
```diff
@@ -91,7 +94,7 @@ export const migrateWorld = async function() {
...
 
   // Set the migration as complete
-  game.settings.set("essence20", "systemMigrationVersion", game.system.version);
+  // game.settings.set("essence20", "systemMigrationVersion", game.system.version);
```
[the remark this answers is not in the document]
Comment (JFranklin3000) [reply module/migration.mjs]: Fixed

**Review by phildominguez — APPROVED** [review]

**Pull request merged**